### PR TITLE
Fix/otel genai

### LINKED ISSUE #26: Incompatibilities with OpenTelemetry LLM semantics pending release
I work on the OpenTelemetry LLM semantics SIG, and did an evaluation of the SDK based on the following sample code and what the semantics pending release 1.27.0 will define.

Note: I'm doing this unsolicited on all the various python instrumentation for openai, so this is not a specific call out that AGIFlow is notably different here. I wanted to warn you about some drift and ideally you'll be in a position to adjust once the [release](https://github.com/open-telemetry/semantic-conventions/releases) occurs, or clarify if that's not a goal. I would welcome you to join the #otel-llm-semconv-wg slack and any [SIG meetings](https://github.com/open-telemetry/community?tab=readme-ov-file#specification-sigs) if you find this relevant!

### Sample code

```python
import os
from agiflow import Agiflow
from openai import OpenAI
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter

# Initialize otel exporter and AGIFlow instrumentation
app_name = "agiflow-python-ollama"
otlp_endpoint = os.getenv("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT", "http://localhost:4318/v1/traces")
otlp_exporter = OTLPSpanExporter(endpoint=otlp_endpoint)
Agiflow.init(app_name=app_name, exporter=otlp_exporter)

def main():
    ollama_host = os.getenv('OLLAMA_HOST', 'localhost')
    # Use the OpenAI endpoint, not the Ollama API.
    base_url = 'http://' + ollama_host + ':11434/v1'
    client = OpenAI(base_url=base_url, api_key='unused')
    messages = [
      {
        'role': 'user',
        'content': '<|fim_prefix|>def hello_world():<|fim_suffix|><|fim_middle|>',
      },
    ]
    chat_completion = client.chat.completions.create(model='codegemma:2b-code', messages=messages)
    print(chat_completion.choices[0].message.content)

if __name__ == "__main__":
    main()
```

### Evaluation

Semantic evaluation on [spans](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-spans.md).

compatible:
* kind=Client

missing:
* attributes['gen_ai.operation.name']='chat'
* attributes['gen_ai.system]='ollama'

incompatible:
* name=Completions (should be 'chat codegemma:2b-code')
* attributes['lllm.type']='Chat' (should be 'gen_ai.operation.name' and lowercase)
* attributes['llm.prompts']='[{"role": "user", "content": "<|fim_prefix|>def hello_world():<|fim_suffix|><|fim_middle|>"}]'  (should be the event attribute 'gen_ai.prompt')
* attributes['llm.model']='codegemma:2b-code' (should be 'gen_ai.request.model')
* attributes['llm.responses']='[{"content": "print(\"Hello, world!\")", "role": "assistant"}]'  (should be the event attribute 'gen_ai.completion')
* attributes['llm.token.counts']='{"prompt_tokens": 24, "completion_tokens": 12, "total_tokens": 36}' (should be split into 'gen_ai.usage.input_tokens' and 'gen_ai.usage.output_tokens')

not yet defined in the standard:
* attributes['openai.api_base']='http://localhost:11434/v1/'
* attributes['llm.api']='/chat/completions'
* attributes['llm.system.fingerprint']='fp_ollama'

defined by other semantics:
* attributes['url.full']='http://localhost:11434/v1/'

vendor specific:
* attributes['agiflow.sdk.name']='agiflow-python-sdk'
* attributes['agiflow.sdk.version']='0.0.23'
* attributes['agiflow.service.name']='OpenAI'
* attributes['agiflow.service.type']='LLM'
* attributes['agiflow.service.version']='1.37.0'

Semantic evaluation on [metrics](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-metrics.md):

N/A as no metrics are currently recorded

### Example collector log

```
otel-collector      | 2024-07-24T05:08:59.563Z  info    TracesExporter  {"kind": "exporter", "data_type": "traces", "name": "debug", "resource spans": 1, "spans": 1}
otel-collector      | 2024-07-24T05:08:59.563Z  info    ResourceSpans #0
otel-collector      | Resource SchemaURL: 
otel-collector      | Resource attributes:
otel-collector      |      -> service.name: Str(agiflow-python-ollama)
otel-collector      |      -> service.version: Str()
otel-collector      |      -> telemetry.sdk.name: Str(AGIFlow)
otel-collector      |      -> telemetry.sdk.version: Str(0.0.23)
otel-collector      | ScopeSpans #0
otel-collector      | ScopeSpans SchemaURL: 
otel-collector      | InstrumentationScope agiflow.opentelemetry.instrumentation.openai.instrumentation 0.0.23
otel-collector      | Span #0
otel-collector      |     Trace ID       : 3d39854a707e30493a3400ba75d0cfc0
otel-collector      |     Parent ID      : 
otel-collector      |     ID             : de037108c788013f
otel-collector      |     Name           : Completions
otel-collector      |     Kind           : Client
otel-collector      |     Start time     : 2024-07-24 05:08:58.820558 +0000 UTC
otel-collector      |     End time       : 2024-07-24 05:08:59.4852 +0000 UTC
otel-collector      |     Status code    : Ok
otel-collector      |     Status message : 
otel-collector      | Attributes:
otel-collector      |      -> agiflow.sdk.name: Str(agiflow-python-sdk)
otel-collector      |      -> agiflow.sdk.version: Str(0.0.23)
otel-collector      |      -> agiflow.service.name: Str(OpenAI)
otel-collector      |      -> agiflow.service.type: Str(LLM)
otel-collector      |      -> agiflow.service.version: Str(1.37.0)
otel-collector      |      -> openai.api_base: Str(http://localhost:11434/v1/)
otel-collector      |      -> url.full: Str(http://localhost:11434/v1/)
otel-collector      |      -> llm.api: Str(/chat/completions)
otel-collector      |      -> llm.type: Str(Chat)
otel-collector      |      -> llm.prompts: Str([{"role": "user", "content": "<|fim_prefix|>def hello_world():<|fim_suffix|><|fim_middle|>"}])
otel-collector      |      -> llm.model: Str(codegemma:2b-code)
otel-collector      |      -> llm.responses: Str([{"content": "print(\"Hello, world!\")", "role": "assistant"}])
otel-collector      |      -> llm.system.fingerprint: Str(fp_ollama)
otel-collector      |      -> llm.token.counts: Str({"prompt_tokens": 24, "completion_tokens": 12, "total_tokens": 36})
otel-collector      |   {"kind": "exporter", "data_type": "traces", "name": "debug"}
```

### PR DESCRIPTION
**Pull request by vuongngo**
Update span attributes to support [Otel GenAi convention](https://opentelemetry.io/docs/specs/semconv/attributes-registry/gen-ai/)

## Add attributes
- gen_ai.response.finish_reasons
- gen_ai.system
- gen_ai.usage.search_units

## Change attributes
- `llm.type` to `gen_ai.operation.name`
- `llm.prompts` to span events with `gen_ai.content.prompt` name and `gen_ai.promp` attributes
- `llm.model` to `gen_ai.request.model` and `gen_ai.response.model`
- `llm.responses` to span events with `gen_ai.content.completion` name and `gen_ai.completion` attributes
- `llm.max_tokens` to `gen_ai.request.max_tokens`
- `llm.temperature` to `gen_ai.request.temperature`
- `llm.top_p` to `gen_ai.request.top_p`
- `llm.response_id` to `gen_ai.response.id`
- `llm.usage.tokens` to `gen_ai.usage.promp_tokens` and `gen_ai.usage.completion_tokens`

## Fix span name
- Change `Chat Completion` span name to `chat [MODEL_NAME]`

Fixes https://github.com/AgiFlow/agiflow-sdks/issues/26

### CODE REVIEW

File: apps/agiflow-sdk-samples/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+``` sh
+poetry shell
+```
+
+## Test scripts
+
+### OpenAI Ollama
+
+``` sh
+python run app/openai_ollama.py
```
Comment (codefromthecrypt) [inline apps/agiflow-sdk-samples/README.md]:
```suggestion
python app/openai_ollama.py
```

**Review by codefromthecrypt — COMMENTED** [review]
Thanks, looks much closer. I understand you are not planning to move to span events, so skip compat on 1.27.0 on the prompt/completion data. That said, there are a couple glitches you might consider as they look easy.

Semantic evaluation on [spans](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-spans.md).

compatible:
* kind=Client
* name='chat codegemma:2b-code'
* attributes['gen_ai.operation.name']='chat'
* attributes['gen_ai.system]='ollama'
* attributes['gen_ai.request.model']='codegemma:2b-code'
* attributes['gen_ai.usage.prompt_tokens']=24

incompatible:
* attributes['genai.prompt']='[{"role": "user", "content": "<|fim_prefix|>def hello_world():<|fim_suffix|><|fim_middle|>"}]'  (should be a span event not an attribute)
* attributes['gen_ai.completion']='[{"content": "print(\"Hello, world!\")", "role": "assistant"}]'   (should be a span event not an attribute)
* attributes['gen_ai.usage.completion_tokens']=13  (should be 'gen_ai.usage.output_tokens')

not yet defined in the standard:
* attributes['openai.api_base']='http://localhost:11434/v1/'
* attributes['llm.api']='/chat/completions'
* attributes['llm.system.fingerprint']='fp_ollama'

defined by other semantics:
* attributes['url.full']='http://localhost:11434/v1/'

vendor specific:
* attributes['agiflow.sdk.name']='agiflow-python-sdk'
* attributes['agiflow.sdk.version']='0.0.23'
* attributes['agiflow.service.name']='OpenAI'
* attributes['agiflow.service.type']='LLM'
* attributes['agiflow.service.version']='1.38.0'

Screenshot 2024-08-05 at 11 04 22 AM

p.s. if you like the screenshot, give it a star ;) https://github.com/ymtdzzz/otel-tui

File: apps/agiflow-sdk-samples/README.md (hunk 1 above)
Comment (vuongngo) [reply apps/agiflow-sdk-samples/README.md]: Thanks @codefromthecrypt  for the details, make my updates easier :). 

Completion and prompt span events are now supported. If you don't have other concern, I'll merge this as soon as our backend support changes in span events.

Screenshot 2024-08-05 at 10 33 32 pm

Comment (vuongngo) [reply apps/agiflow-sdk-samples/README.md]: Thanks, this has been updated!

**Comment by vuongngo** [comment]
Thanks @codefromthecrypt for the details, make my updates easier :).

Completion and prompt span events are now supported. If you don't have other concern, I'll merge this as soon as our backend support changes in span events.

Screenshot 2024-08-05 at 10 33 32 pm

**Comment by codefromthecrypt** [comment]
cool. 100pct compatible!

Screenshot 2024-08-06 at 9 52 19 AM

**Comment by vuongngo** [comment]
Awesome, thank you! @codefromthecrypt

**Pull request merged**